File: IrrgartenRuby/modelo/dice.rb
# frozen_string_literal: true
module Modelo
  class Dice
    @@MAX_USES = 5
    @@MAX_INTELLIGENCE = 10.0
    @@MAX_STRENGTH = 10.0
    @@RESURRECT_PROB = 0.3
    @@WEAPONS_REWARD = 2
    @@SHIELDS_REWARD = 3
    @@HEALTH_REWARD = 5
    @@MAX_ATTACK = 3
    @@MAX_SHIELD = 2


    @@generador = Random.new


  #Ejemplo de uso del generador de clase Random
    def self.aleatorio
      if @@generador.rand < 0.3
        return true
      else
        return false
      end

    end

  #Devuelve un número de fila o columna aleatoria
    def self.random_pos(max)
      @@generador.rand(0..max)
    end

    def self.who_starts(nplayer)
      @@generador.rand(0..nplayer)
    end

    def self.random_intelligence
      #@@generador.rand(0...@@MAX_INTELLIGENCE) TRES PUNTOS
      @@generador.rand(@@MAX_INTELLIGENCE.to_f)
    end

    def self.random_strenght
      @@generador.rand(@@MAX_STRENGTH.to_f)
    end

    def self.resurrect_player
      @@generador.rand() > @@RESURRECT_PROB
    end

    def self.weapons_reward
      @@generador.rand(0..@@WEAPONS_REWARD)
    end

    def self.shield_reward
      @@generador.rand(0..@@SHIELDS_REWARD)
    end

    def self.health_reward
      @@generador.rand(0..@@HEALTH_REWARD)
    end

    def self.weapon_power
      @@generador.rand(0...@@MAX_ATTACK.to_f)
    end

    def self.shield_power
      @@generador.rand(0...@@MAX_SHIELD.to_f)
    end

    def self.uses_left
      @@generador.rand(0..@@MAX_USES)
    end

    def self.intensity(competence)
      @@generador.rand(0...competence)
    end

    def self.discard_element(uses_left)
      probability = 1.to_f/(uses_left.to_f / @@MAX_USES)
      return @@generador.rand() < probability
    end

  end
end

File: IrrgartenJava/src/modelo/Dice.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package modelo;

import java.util.Random;

/**
 *
 * @author nmatl
 */
public class Dice {
    static private final int MAX_USES = 5;
    static private final float MAX_INTELLIGENCE = 10.0f;
    static private final float MAX_STRENGTH = 10.0f;  
    static private final float RESURRECT_PROB = 0.3f;
    static private final int WEAPONS_REWARD = 2; 
    static private final int SHIELDS_REWARD = 3; 
    static private final int HEALTH_REWARD = 5;
    static private final int MAX_ATTACK = 3; 
    static private final int MAX_SHIELD = 2; 
    
    static private Random generador = new Random();
    
    static public int randomPos(int max){
        return generador.nextInt(max + 1);
    }
    
    static public int whoStarts(int nplayers){
        return generador.nextInt(nplayers + 1);
    }
    
    static public float randomIntelligence(){
     // nextFloat() genera [0.0, 1.0), al multiplicar por max obtenemos [0.0, max)
         return generador.nextFloat() * MAX_INTELLIGENCE;
    }
    
    static public boolean resurrectPlayer(){
        return generador.nextFloat() > RESURRECT_PROB;
    }
    
    static public int weaponsReward(){
        return generador.nextInt(WEAPONS_REWARD + 1);
    }
    
    static public int shieldsReward(){
        return generador.nextInt(SHIELDS_REWARD + 1);
    }
    
    static public int healthReward(){
        return generador.nextInt(HEALTH_REWARD + 1);
    }
    
    static public float weaponPower(){
        return generador.nextFloat() * MAX_ATTACK;
    }
    
    static public float shieldPower(){
        return generador.nextFloat() * MAX_SHIELD;
    }
  
    static public int usesLeft(){
        return generador.nextInt(MAX_USES + 1);
    }
     static public float intensity(float competence){
         return generador.nextFloat() * competence;
     }
     
     static public boolean discardElement(int usesLeft){
         float probability = 1.0f - ((float)usesLeft / MAX_USES);
         return generador.nextFloat() < probability;
     }
    
}
